src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -0,0 +1,235 @@
+      final Character delimiter) {
+    this(source, schema, workerID, totalWorkers, delimiter, null, null, null);
+  }
+
+  public CSVFileScanFragment(final String filename, final Schema schema, final int workerID, final int totalWorkers,
+      @Nullable final Character delimiter, @Nullable final Character quote, @Nullable final Character escape,
+      @Nullable final Integer numberOfSkippedLines) {
+    this(new FileSource(filename), schema, workerID, totalWorkers, delimiter, quote, escape, numberOfSkippedLines);
+  }
+
+  public CSVFileScanFragment(final DataSource source, final Schema schema, final int workerID, final int totalWorkers,
+      @Nullable final Character delimiter, @Nullable final Character quote, @Nullable final Character escape,
+      @Nullable final Integer numberOfSkippedLines) {
+    this.source = (AmazonS3Source) Preconditions.checkNotNull(source, "source");
+    this.schema = Preconditions.checkNotNull(schema, "schema");
+
+    this.delimiter = MoreObjects.firstNonNull(delimiter, CSVFormat.DEFAULT.getDelimiter());
+    this.quote = MoreObjects.firstNonNull(quote, CSVFormat.DEFAULT.getQuoteCharacter());
+    this.escape = escape != null ? escape : CSVFormat.DEFAULT.getEscapeCharacter();
+    this.numberOfSkippedLines = MoreObjects.firstNonNull(numberOfSkippedLines, 0);
+    this.workerID = workerID;
+
+    isLastWorker = workerID == totalWorkers;
+    fileSize = ((AmazonS3Source) source).getFileSize();
+
+    if (fileSize < 10 && workerID == 1) {
+      startByteRange = 0;
+      endByteRange = fileSize;
+    } else {
+      long partitionSize = fileSize / totalWorkers;
+      startByteRange = (partitionSize) * (workerID - 1);
+      endByteRange = startByteRange + partitionSize;
+    }
+
+  }
+
+  @Override
+  protected TupleBatch fetchNextReady() throws IOException, DbException {
+    long lineNumberBegin = lineNumber;
+    boolean fixingStartByte = false;
+    boolean onLastRow = false;
+
+    while ((buffer.numTuples() < TupleBatch.BATCH_SIZE)) {
+      lineNumber++;
+      if (parser.isClosed()) {
+        break;
+      }
+      try {
+        if (!iterator.hasNext()) {
+          parser.close();
+          break;
+        }
+      } catch (final RuntimeException e) {
+        throw new DbException("Error parsing row " + lineNumber, e);
+      }
+      CSVRecord record = iterator.next();
+      if (record.size() < schema.numColumns()) {
+        if (lineNumber - 1 != 0 && !isLastWorker) {
+          onLastRow = true;
+          long byteAtBeginningOfRecord = record.getCharacterPosition();
+          if (!fixingStartByte) {
+            fixingStartByte = true;
+            startByteRange += byteAtBeginningOfRecord;
+          }
+          byteOverlap = (long) Math.pow(byteOverlap, 2);
+          endByteRange = endByteRange + byteOverlap;
+
+          source.setStartRange(startByteRange);
+          source.setEndRange(endByteRange);
+          parser =
```

Comment:
If we concatenate successive overlap ranges using the approach I suggested, we should be able to avoid reinitializing the parser (its position in the `SequenceInputStream` would remain the same).